### PR TITLE
QgisApp::addVectorLayer(): add |layername= to OGR datasets (fixes #20031)

### PR DESCRIPTION
**Pull request by rouault**
Do it also in case of datasets that have a single layer, in case they might
later be edited to have more layers
(except for a few drivers known to be always single layer)

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -4441,6 +4441,83 @@ QString QgisApp::crsAndFormatAdjustedLayerUri( const QString &uri, const QString
...
+    QString uri = layer->source();
+    QString name = layer->name();
+    //the separator char & was changed to | to be compatible
+    //with url for protocol drivers
+    if ( uri.contains( '|', Qt::CaseSensitive ) )
+    {
+      // If we get here, there are some options added to the filename.
+      // A valid uri is of the form: filename&option1=value1&option2=value2,...
+      // We want only the filename here, so we get the first part of the split.
+      QStringList theURIParts = uri.split( '|' );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Candidate for QgsDataProvider::decodeUri?

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -4441,6 +4441,83 @@ QString QgisApp::crsAndFormatAdjustedLayerUri( const QString &uri, const QString
...
+      uri = theURIParts.at( 0 );
+    }
+
+    QString composedURI = uri;
+    composedURI += "|layername=" + rawLayerName;
+
+    QgsVectorLayer *newLayer = new QgsVectorLayer( composedURI, name, QStringLiteral( "ogr" ), options );
+    if ( newLayer && newLayer->isValid() )
+    {
+      delete layer;
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Looks like some unique_ptr candidates here

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (rouault) [reply src/app/qgisapp.cpp]: Done

File: src/app/qgisapp.cpp (hunk 2 above)
Comment (rouault) [reply src/app/qgisapp.cpp]: Done

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**